### PR TITLE
[SPARK-40663][SQL](Final) Migrate execution errors onto error classes

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This the final PR that proposes to migrate the execution errors onto temporary error classes with the prefix `_LEGACY_ERROR_TEMP`

The error classes are prefixed with `_LEGACY_ERROR_TEMP_` indicates the dev-facing error messages, and won't be exposed to end users.

### Why are the changes needed?

To speed-up the error class migration.

The migration on temporary error classes allow us to analyze the errors, so we can detect the most popular error classes.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

```
$ build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite"
$ build/sbt "test:testOnly *SQLQuerySuite"
$ build/sbt -Phive-thriftserver "hive-thriftserver/testOnly org.apache.spark.sql.hive.thriftserver.ThriftServerQueryTestSuite"
```

### CODE REVIEW

**Comment by itholic** [comment]
This is ready for review, @MaxGekk FYI

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Comment by itholic** [comment]
Thanks for the review, @MaxGekk !

Just adjusted the comments

**Comment by MaxGekk** [comment]
@itholic Could you re-trigger GAs and remove `[WIP]` if this PR is ready.

**Comment by itholic** [comment]
Thanks! Just re-triggered the GA and fix the title!

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
@itholic Please, fix the test failure:
```
[info] - postgreSQL/numeric.sql *** FAILED *** (49 seconds, 862 milliseconds)
[info]   postgreSQL/numeric.sql
[info]   Expected "...g.apache.spark.Spark[IllegalArgumentException
[info]   {
[info]     "errorClass" : "_LEGACY_ERROR_TEMP_2278",
[info]     "messageParameters" : {
[info]       "format" : "99G999G999",
[info]       "input" : "-34,338,492]"
[info]     }
[info]   }", but got "...g.apache.spark.Spark[Exception
[info]   {
[info]     "errorClass" : "INTERNAL_ERROR",
[info]     "messageParameters" : {
[info]       "message" : "Undefined error message parameter for error class: '_LEGACY_ERROR_TEMP_2278'. Parameters: Map(input -> -34,338,492, format -> 99G999G999)]"
```

**Comment by MaxGekk** [comment]
ping @itholic

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala [hunk 1]
```diff
@@ -1124,8 +1124,9 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     ).foreach { case (str: String, format: String) =>
       val toNumberExpr = ToNumber(Literal(str), Literal(format))
       assert(toNumberExpr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
-      checkExceptionInExpression[IllegalArgumentException](
-        toNumberExpr, "does not match the given number format")
+
+      checkExceptionInExpression[SparkException](
+        toNumberExpr, "[INTERNAL_ERROR] Undefined error message parameter")
```
Comment (itholic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1129]: Sorry for late responding, @MaxGekk .

I've been thinking about if there is better way to test expression with error class here.

Please let me know if there is better way to test this ? 🙏

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -2602,31 +2604,46 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase {
...
-  def invalidNumberFormatError(valueType: String, input: String, format: String): Throwable = {
-    new IllegalArgumentException(
-      s"The input $valueType '$input' does not match the given number format: '$format'")
+  def invalidNumberFormatError(
+      valueType: String, input: String, format: String): SparkIllegalArgumentException = {
+    new SparkIllegalArgumentException(
+      errorClass = "_LEGACY_ERROR_TEMP_2278",
+      messageParameters = Map(
+        "input" -> input,
+        "format" -> format))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2622]: Missed `valueType`, see the format of the error class `_LEGACY_ERROR_TEMP_2278`.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1129]: See https://github.com/apache/spark/pull/38177#pullrequestreview-1153416983

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala (hunk 2 above)
Comment (itholic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:2623]: Oops, I missed it... Thanks for figuring out!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala (hunk 1 above)
Comment (itholic) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala:1129]: Seems like I should also fix the test & generated golden files.

Let me fix soon!

**Comment by itholic** [comment]
Fixed & re-gened the golden files!

**Comment by MaxGekk** [comment]
+1, LGTM. All GAs passed. Merging to master.
Thank you, @itholic.

**Pull request closed without merging**